src/wazuh_modules/wm_ciscat.c: support both CIS-CAT Pro V3 and V4

hustliyilin opened the pull request:
Support both CIS-CAT Pro V3 and V4

Signed-off-by: YiLin.Li <YiLin.Li@linux.alibaba.com>

|Related issue|
|---|
| #12581 |

<!--
This template reflects sections that must be included in new Pull requests.
Contributions from the community are really appreciated. If this is the case, please add the
"contribution" to properly track the Pull Request.

Please fill the table above. Feel free to extend it at your convenience.
-->

## Description

<!--
Add a clear description of how the problem has been solved.
-->

## Configuration options

<!--
When proceed, this section should include new configuration parameters.
-->

## Logs/Alerts example

<!--
Paste here related logs and alerts
-->

## Tests

<!--
Depending on the affected components by this PR, the following checks should be selected and marked.
-->

<!-- Minimum checks required -->
- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
  - [ ] MAC OS X
- [ ] Source installation
- [ ] Package installation
- [ ] Source upgrade
- [ ] Package upgrade
- [ ] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities

<!-- Depending on the affected OS -->
- Memory tests for Linux
  - [ ] Scan-build report
  - [ ] Coverity
  - [ ] Valgrind (memcheck and descriptor leaks check)
  - [ ] Dr. Memory
  - [ ] AddressSanitizer
- Memory tests for Windows
  - [ ] Scan-build report
  - [ ] Coverity
  - [ ] Dr. Memory
- Memory tests for macOS
  - [ ] Scan-build report
  - [ ] Leaks
  - [ ] AddressSanitizer

<!-- Checks for huge PRs that affect the product more generally -->
- [ ] Retrocompatibility with older Wazuh versions
- [ ] Working on cluster environments
- [ ] Configuration on demand reports new parameters
- [ ] The data flow works as expected (agent-manager-api-app)
- [ ] Added unit tests (for new features)
- [ ] Stress test for affected components

<!-- Ruleset required checks, rules/decoder -->
- Decoder/Rule tests
  - [ ] Added unit testing files ".ini"
  - [ ] runtests.py executed without errors

vikman90 reviewed (changes requested):
@hustliyilin Thank you for opening this PR, I appreciate it.

However, I'm afraid this change would break current installations with a legacy version of CIS-CAT. So, we would need:

- A new option like `<ciscat_binary>` to allow the user to change between `CIS-CAT.sh` and `Assessor-CLI.sh`. The default value must be the former, until a new major version.
- Apply the same change on the Windows agent, which currently runs `CIS-CAT.BAT`.
- Add the new option to the documentation.
- Test the module against CIS-CAT Pro Assessor v4 and check that it works properly. Attach evidence in this PR.

These changes are essential for the agent to continue working with previous versions of CIS-CAT.

Thanks again.
Best regards.

hustliyilin commented:
> @hustliyilin Thank you for opening this PR, I appreciate it.
> 
> However, I'm afraid this change would break current installations with a legacy version of CIS-CAT. So, we would need:
> 
> * A new option like `<ciscat_binary>` to allow the user to change between `CIS-CAT.sh` and `Assessor-CLI.sh`. The default value must be the former, until a new major version.
> * Apply the same change on the Windows agent, which currently runs `CIS-CAT.BAT`.
> * Add the new option to the documentation.
> * Test the module against CIS-CAT Pro Assessor v4 and check that it works properly. Attach evidence in this PR.
> 
> These changes are essential for the agent to continue working with previous versions of CIS-CAT.
> 
> Thanks again. Best regards.

Hi @vikman90 , 

Thanks for your review suggestion. Now 
- I add the new configuration `<ciscat_binary>`  to allow the users to change between `CIS-CAT`.sh and `Assessor-CLI.sh`. The default value is the former.
- I support both CIS-CAT Pro V3 and V4 in UNIX systems and push all commits.
- The test results in CentOS 7.9 are the following

# CIS-CAT Pro V3
## Without the new configuration `<ciscat_binary>` in the `/var/ossec/etc/ossec.conf`

```XML
  <wodle name="cis-cat">
    <disabled>no</disabled>
    <timeout>1800</timeout>
    <interval>5m</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.322.b06-1.el7_9.x86_64/jre/bin</java_path>
    <ciscat_path>/root/cis-cat-full</ciscat_path>

    <content type="xccdf" path="benchmarks/CIS_CentOS_Linux_7_Benchmark_v3.1.2-xccdf.xml">
       <profile>"xccdf_org.cisecurity.benchmarks_profile_Level_2_-_Server"</profile>
    </content>

  </wodle>
```

and the test results (RESTful API and ossec.log) are

<img width="1157" alt="lQLPDhtALHIl2Q_NA3TNCQqw7A8OsZIdI5wCOdC0KkBsAA_2314_884" src="https://user-images.githubusercontent.com/6824093/159112347-682069c5-0eca-480e-a23d-f8f41e8ceeea.png">

## With the new configuration `<ciscat_binary>CIS-CAT.sh</ciscat_binary>` in the `/var/ossec/etc/ossec.conf`

```XML
  <wodle name="cis-cat">
    <disabled>no</disabled>
    <timeout>1800</timeout>
    <interval>5m</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.322.b06-1.el7_9.x86_64/jre/bin</java_path>
    <ciscat_path>/root/cis-cat-full</ciscat_path>
    <ciscat_binary>CIS-CAT.sh</ciscat_binary>

    <content type="xccdf" path="benchmarks/CIS_CentOS_Linux_7_Benchmark_v3.1.2-xccdf.xml">
       <profile>"xccdf_org.cisecurity.benchmarks_profile_Level_2_-_Server"</profile>
    </content>

  </wodle>
```

and the test results (RESTful API and ossec.log) are

<img width="1157" alt="lQLPDhtALHIl2Q_NA3TNCQqw7A8OsZIdI5wCOdC0KkBsAA_2314_884" src="https://user-images.githubusercontent.com/6824093/159112418-d17dbc31-9030-4b21-8d7b-f43adc3c50e1.png">

# CIS-CAT Pro V4
## with the new configuration `<ciscat_binary>Assessor-CLI.sh</ciscat_binary>` in the `/var/ossec/etc/ossec.conf`

```XML
  <wodle name="cis-cat">
    <disabled>no</disabled>
    <timeout>1800</timeout>
    <interval>5m</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.322.b06-1.el7_9.x86_64/jre/bin</java_path>
    <ciscat_path>/root/Assessor-CLI</ciscat_path>
    <ciscat_binary>Assessor-CLI.sh</ciscat_binary>

    <content type="xccdf" path="benchmarks/CIS_CentOS_Linux_7_Benchmark_v3.1.2-xccdf.xml">
       <profile>"Level 2 - Server"</profile>
    </content>

  </wodle>
```

and the test results (RESTful API and ossec.log) are

<img width="1157" alt="lQLPDhtAKxkJd6PNA5zNCQqwdTUQTRbEB9gCOc5_MIBTAA_2314_924" src="https://user-images.githubusercontent.com/6824093/159112480-0d8aad1d-347f-4e55-972f-fd62f3ca057e.png">

Now I can support both CIS-CAT Pro V3 and V4 in UNIX systems with wazuh.  I would really appreciate if you could review these codes. 

However, There are still two questions to be solved in the next weeks.
- Support Both CIS-CAT Pro V3 and V4 in windows: **I am a new windows developer, Could you provide me some documents about how to build wazuh from source codes in Windows?**
- Update documentation:  How Can I Add the new option to the documentation? 
Thanks again @vikman90.  I look forward to your reply.

vikman90 commented on src/wazuh_modules/wm_ciscat.c:
```diff
@@ -167,6 +167,26 @@ void* wm_ciscat_main(wm_ciscat *ciscat) {
     }
 
+    #ifdef WIN32
+        if (ciscat->ciscat_binary == '\0') {
```
`ciscat_binary` is `char *`. Thus, we should compare it with a pointer.
```suggestion
        if (ciscat->ciscat_binary == NULL) {
```

hustliyilin replied:
> `ciscat_binary` is `char *`. Thus, we should compare it with a pointer.
> 
> ```
> wazuh_modules/wm_ciscat.c:180:35: warning: comparison between pointer and zero character constant [-Wpointer-compare]
> ```

If we change to compare it with a pointer. The previous user needs to add the configuration  `<ciscat_binary>` in the `/var/ossec/etc/ossec.conf` manully, otherwise there is an error like this

![image](https://user-images.githubusercontent.com/6824093/160048135-3e932c9a-9c1e-43a9-9aca-933ec456aaad.png)


If we compare it with `\0`,  The above error will no longer appear. @vikman90

vikman90 commented on src/wazuh_modules/wm_ciscat.c:
```diff
@@ -167,6 +167,26 @@ void* wm_ciscat_main(wm_ciscat *ciscat) {
+            ciscat->flags.error = 1;
+        }
+    #else
+        if (ciscat->ciscat_binary == '\0') {
```
Same issue here.
```suggestion
        if (ciscat->ciscat_binary == NULL) {
```

hustliyilin replied:
> Same issue here.

`ciscat->ciscat_binary` is a string variable. We need compare `\0` instead of `NULL` to judge whether the string is empty or not. @vikman90

vikman90 commented on src/wazuh_modules/wm_ciscat.c:
```diff
@@ -167,6 +167,26 @@ void* wm_ciscat_main(wm_ciscat *ciscat) {
+            ciscat->flags.error = 1;
+        }
+    #else
+        if (ciscat->ciscat_binary == '\0') {
```
Same issue here.
```suggestion
        if (ciscat->ciscat_binary == NULL) {
```
```suggestion
        if (ciscat->ciscat_binary == '\0') {
```

vikman90 commented on src/wazuh_modules/wm_ciscat.c:
```diff
@@ -167,6 +167,26 @@ void* wm_ciscat_main(wm_ciscat *ciscat) {
+            ciscat->ciscat_binary = WM_CISCAT_V3_BINARY;
+        }
+
+        if (strcmp(ciscat->ciscat_binary, WM_CISCAT_V3_BINARY) && strcmp(ciscat->ciscat_binary, WM_CISCAT_V4_BINARY)) {
```
This code here is checking the configuration. Why not run it at _src/config/wmodules-ciscat.c_?

vikman90 replied:
This code here is checking the configuration. Why not run it at _src/config/wmodules-ciscat.c_?

vikman90 commented on src/wazuh_modules/wm_ciscat.c:
```diff
@@ -284,7 +304,7 @@ void wm_ciscat_cleanup() {
 #ifdef WIN32
 
-void wm_ciscat_run(wm_ciscat_eval *eval, char *path, int id, const char * java_path) {
+void wm_ciscat_run(wm_ciscat_eval *eval, char *path, int id, const char *java_path, const char *ciscat_binary) {
```
`ciscat_binary` seems not to be used here. It still hardcodes `CIS-CAT.BAT`.

vikman90 replied:
`ciscat_binary` seems not to be used here. It still hardcodes `CIS-CAT.BAT`.

vikman90 commented on src/wazuh_modules/wm_ciscat.c:
```diff
@@ -846,11 +902,11 @@ wm_scan_data* wm_ciscat_txt_parser(){
         }
 
-        mtdebug1(WM_CISCAT_LOGTAG, "Finished parse of the TXT report.");
+        mtinfo(WM_CISCAT_LOGTAG, "Finished parse of the TXT report.");
```
I think this log is too verbose to appear as "INFO".
```
DEBUG: Launching command: ./CIS-CAT.sh -a -b /var/ossec/wodles/cis-cat/benchmarks/CIS_CentOS_Linux_7_Benchmark_v2.1.1-xccdf.xml -p "xccdf_org.cisecurity.benchmarks_profile_Level_2_-_Server" -r /var/ossec/tmp -rn ciscat-report -x -t -n -y
```

vikman90 replied:
I think this log is too verbose to appear as "INFO".
```
DEBUG: Launching command: ./CIS-CAT.sh -a -b /var/ossec/wodles/cis-cat/benchmarks/CIS_CentOS_Linux_7_Benchmark_v2.1.1-xccdf.xml -p "xccdf_org.cisecurity.benchmarks_profile_Level_2_-_Server" -r /var/ossec/tmp -rn ciscat-report -x -t -n -y
```

vikman90 commented on src/wazuh_modules/wm_ciscat.c:
```diff
@@ -448,12 +468,11 @@ void wm_ciscat_run(wm_ciscat_eval *eval, char *path, int id, const char * java_p
-    // Accepting Terms of Use
-
-    wm_strcat(&command, "-a", ' ');
+    strcpy(tmp, path);
+    strcat(tmp, "/");
+    
+    wm_strcat(&command, tmp, '/');    
+    wm_strcat(&command, ciscat_binary, '\0');
```
This is an insecure operation as we're copying `path`, with an undefined size, into an `OS_MAXSTR` bytes string.
```suggestion
    wm_strcat(&command, path, '/');
    wm_strcat(&command, ciscat_binary, '/');
```

vikman90 replied:
This is an insecure operation as we're copying `path`, with an undefined size, into an `OS_MAXSTR` bytes string.
```suggestion
    wm_strcat(&command, path, '/');
    wm_strcat(&command, ciscat_binary, '/');
```

vikman90 reviewed (changes requested):
@hustliyilin thank you so much for your changes!

I've left some notes that I think are important to apply or discuss.

On the other hand, I miss these things:
- Apply the same changes on Windows.
- Add the new option to the JSON settings dump (`wm_ciscat_dump()`):
```c
if (ciscat->ciscat_binary) cJSON_AddStringToObject(wm_cscat,"ciscat_binary",ciscat->ciscat_binary);
```
- Add this new option to the documentation.

I would be very grateful if you could apply these changes. Otherwise, we can merge this PR into a development branch and change this stuff by ourselves.

On the other hand, I've ran Modulesd on Valgrind, and it reported no memory problems:

```
==31177== LEAK SUMMARY:
==31177==    definitely lost: 0 bytes in 0 blocks
==31177==    indirectly lost: 0 bytes in 0 blocks
==31177==      possibly lost: 5,216 bytes in 9 blocks
==31177==    still reachable: 783,143 bytes in 13,344 blocks
==31177==         suppressed: 0 bytes in 0 blocks
```

Cheers!

hustliyilin commented:
> @hustliyilin thank you so much for your changes!
> 
> I've left some notes that I think are important to apply or discuss.
> 
> On the other hand, I miss these things:
> 
> * Apply the same changes on Windows.
> * Add the new option to the JSON settings dump (`wm_ciscat_dump()`):
> 
> ```c
> if (ciscat->ciscat_binary) cJSON_AddStringToObject(wm_cscat,"ciscat_binary",ciscat->ciscat_binary);
> ```
> 
> * Add this new option to the documentation.
> 
> I would be very grateful if you could apply these changes. Otherwise, we can merge this PR into a development branch and change this stuff by ourselves.
> 
> On the other hand, I've ran Modulesd on Valgrind, and it reported no memory problems:
> 
> ```
> ==31177== LEAK SUMMARY:
> ==31177==    definitely lost: 0 bytes in 0 blocks
> ==31177==    indirectly lost: 0 bytes in 0 blocks
> ==31177==      possibly lost: 5,216 bytes in 9 blocks
> ==31177==    still reachable: 783,143 bytes in 13,344 blocks
> ==31177==         suppressed: 0 bytes in 0 blocks
> ```
> 
> Cheers!

Hi @vikman90,

Thanks for your review. I can accept almost all suggestions. There are two places where I think your suggestions may not be so correct. I added the reply in the corresponding position. 

In addition, I will also 
- Apply the same changes on Windows.
- Add the new option to the JSON settings dump (wm_ciscat_dump()):
- Add this new option to the documentation.

I will submit the latest version of these codes as soon as possible.

Thanks again.

qrow2022 commented:
@hustliyilin 
I was originally working on this but got distracted. If any of my work would be of help to you feel free to copy it. I've got a fork on my repo here: https://github.com/qrow2022/wazuh/tree/update-CISCat-v4

hustliyilin commented:
> @hustliyilin I was originally working on this but got distracted. If any of my work would be of help to you feel free to copy it. I've got a fork on my repo here: https://github.com/qrow2022/wazuh/tree/update-CISCat-v4 The issue was #6681

Hi @qrow2022,

Thank you very much. I will refer to your work to support both CIS-CAT v3 and v4 in Unix and Windows.  I will add my thanks to you in the commit log at the end. Thanks again.

qrow2022 commented:
@hustliyilin 
I'm not entirely sure what you mean by a 'Signed-off-by'. I assume you mean like this and here in this thread.

I give you permission to use my code as long as there is attribution.

Signed-Off-By: Nathaniel Root <nathanielroot15@gmail.com>

hustliyilin commented:
> > @hustliyilin thank you so much for your changes!
> > I've left some notes that I think are important to apply or discuss.
> > On the other hand, I miss these things:
> > 
> > * Apply the same changes on Windows.
> > * Add the new option to the JSON settings dump (`wm_ciscat_dump()`):
> > 
> > ```c
> > if (ciscat->ciscat_binary) cJSON_AddStringToObject(wm_cscat,"ciscat_binary",ciscat->ciscat_binary);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * Add this new option to the documentation.
> > 
> > I would be very grateful if you could apply these changes. Otherwise, we can merge this PR into a development branch and change this stuff by ourselves.
> > On the other hand, I've ran Modulesd on Valgrind, and it reported no memory problems:
> > ```
> > ==31177== LEAK SUMMARY:
> > ==31177==    definitely lost: 0 bytes in 0 blocks
> > ==31177==    indirectly lost: 0 bytes in 0 blocks
> > ==31177==      possibly lost: 5,216 bytes in 9 blocks
> > ==31177==    still reachable: 783,143 bytes in 13,344 blocks
> > ==31177==         suppressed: 0 bytes in 0 blocks
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Cheers!
> 
> Hi @vikman90,
> 
> Thanks for your review. I can accept all suggestions.
> 
> In addition, I will also
> 
> * Apply the same changes on Windows.
> * Add the new option to the JSON settings dump (wm_ciscat_dump()):
> * Add this new option to the documentation.
> 
> I will submit the latest version of these codes as soon as possible.
> 
> Thanks again.

Hi again @vikman90,

Now I added this new option to the documentation in https://github.com/wazuh/wazuh-documentation/pull/4976 and added the new option to the JSON settings dump (wm_ciscat_dump()) in https://github.com/wazuh/wazuh/pull/12582/files#diff-91a6d96078120b0454e751eb828ee4c5c301e7bfea10b866b22a84bd3fe62d2eR1546 .

Besides, I applied the same changes on Windows and modified the codes as suggested by your review. The test results are the following now.

# Windows 10 Agent

The agent `001` is `Windows 10`

<img width="1157" alt="lALPJwnIxvvZxC3NBMvNCQo_2314_1227" src="https://user-images.githubusercontent.com/6824093/160772857-af8f346f-90bb-4846-a79d-89f85da35031.png">

## CIS-CAT Pro v3
### Without the new configuration `<ciscat_binary>` in the `ossec.conf`

```XML
  <!-- CIS policies evaluation -->
  <wodle name="cis-cat">
    <disabled>no</disabled>
    <timeout>1800</timeout>
    <interval>8m</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>C:\Program Files\Java\jre1.8.0_311\bin</java_path>
    <ciscat_path>D:\test\ciscat-full-bundle-2021-12-20-v3.0.78\cis-cat-full</ciscat_path>

    <content type="xccdf" path="D:\test\ciscat-full-bundle-2021-12-20-v3.0.78\cis-cat-full\benchmarks\CIS_Microsoft_Windows_10_Enterprise_Release_21H1_Benchmark_v1.11.0-xccdf.xml">
           <profile>"Level 1 (L1) - Corporate/Enterprise Environment (general use)"</profile>
     </content>
  </wodle>
```

and the test results (RESTful API and ossec.log) are

<img width="1155" alt="lALPJx8Zvth-qV3NAnXNCQY_2310_629" src="https://user-images.githubusercontent.com/6824093/160773005-6d70ed68-6f50-4301-a185-4da7bf80a935.png">


### With the new configuration `<ciscat_binary>CIS-CAT.BAT</ciscat_binary>` in the `ossec.conf`

```XML
 <!-- CIS policies evaluation -->
  <wodle name="cis-cat">
    <disabled>no</disabled>
    <timeout>1800</timeout>
    <interval>8m</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>C:\Program Files\Java\jre1.8.0_311\bin</java_path>
    <ciscat_path>D:\test\ciscat-full-bundle-2021-12-20-v3.0.78\cis-cat-full</ciscat_path>
    <ciscat_binary>CIS-CAT.BAT</ciscat_binary>

    <content type="xccdf" path="D:\test\ciscat-full-bundle-2021-12-20-v3.0.78\cis-cat-full\benchmarks\CIS_Microsoft_Windows_10_Enterprise_Release_21H1_Benchmark_v1.11.0-xccdf.xml">
           <profile>"Level 1 (L1) - Corporate/Enterprise Environment (general use)"</profile>
     </content>
  </wodle>
```

and the test results (RESTful API and ossec.log) are

<img width="1157" alt="lALPJxf-wY9PDpvNAnfNCQo_2314_631" src="https://user-images.githubusercontent.com/6824093/160772803-7d67406e-fd4a-4389-a98e-5ea07b823aa7.png">


## CIS-CAT Pro v4
### With the new configuration `<ciscat_binary>Assessor-CLI.bat</ciscat_binary>` in the `ossec.conf`

```XML
 <!-- CIS policies evaluation -->
  <wodle name="cis-cat">
    <disabled>no</disabled>
    <timeout>1800</timeout>
    <interval>8m</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>C:\Program Files\Java\jre1.8.0_311\bin</java_path>
    <ciscat_path>D:\test\Assessor-v4.15.0\Assessor-CLI</ciscat_path>
    <ciscat_binary>Assessor-CLI.bat</ciscat_binary>

    <content type="xccdf" path="D:\test\Assessor-v4.15.0\Assessor-CLI\benchmarks\CIS_Microsoft_Intune_for_Windows_10_Release_2004_Benchmark_v1.0.1-xccdf.xml">
           <profile>"Level 1 (L1) - Corporate/Enterprise Environment (general use)"</profile>
     </content>
  </wodle>
```

and the test results (RESTful API and ossec.log) are

<img width="1155" alt="lALPJw1WRaE8_HrNAnfNCQU_2309_631" src="https://user-images.githubusercontent.com/6824093/160772708-365bf328-8153-4082-a41b-52bbdc233d05.png">


# Unix
## CIS-CAT Pro v3
### Without the new configuration `<ciscat_binary>` in the `/var/ossec/etc/ossec.conf`

```XML
  <wodle name="cis-cat">
    <disabled>no</disabled>
    <timeout>1800</timeout>
    <interval>5m</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.322.b06-1.el7_9.x86_64/jre/bin</java_path>
    <ciscat_path>/root/cis-cat-full</ciscat_path>

    <content type="xccdf" path="benchmarks/CIS_CentOS_Linux_7_Benchmark_v3.1.2-xccdf.xml">
       <profile>"xccdf_org.cisecurity.benchmarks_profile_Level_2_-_Server"</profile>
    </content>

  </wodle>
```

and the test results (RESTful API and ossec.log) are

<img width="1162" alt="lALPJx8ZvtOUM1bNA6_NCRM_2323_943" src="https://user-images.githubusercontent.com/6824093/160771971-b25a6c3a-b1e6-4551-8183-530ff433175f.png">


### With the new configuration `<ciscat_binary>CIS-CAT.sh</ciscat_binary>` in the `/var/ossec/etc/ossec.conf`

```XML
  <wodle name="cis-cat">
    <disabled>no</disabled>
    <timeout>1800</timeout>
    <interval>5m</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.322.b06-1.el7_9.x86_64/jre/bin</java_path>
    <ciscat_path>/root/cis-cat-full</ciscat_path>
    <ciscat_binary>CIS-CAT.sh</ciscat_binary>

    <content type="xccdf" path="benchmarks/CIS_CentOS_Linux_7_Benchmark_v3.1.2-xccdf.xml">
       <profile>"xccdf_org.cisecurity.benchmarks_profile_Level_2_-_Server"</profile>
    </content>

  </wodle>
```

and the test results (RESTful API and ossec.log) are

<img width="1159" alt="lALPJxf-wYpxgDLNA8fNCQ4_2318_967" src="https://user-images.githubusercontent.com/6824093/160772051-c2256b43-0d3b-4fbf-a3c3-4ad1d8eb4d6f.png">


## CIS-CAT Pro v4
### With the new configuration <ciscat_binary>Assessor-CLI.sh</ciscat_binary> in the /var/ossec/etc/ossec.conf

```XML
  <wodle name="cis-cat">
    <disabled>no</disabled>
    <timeout>1800</timeout>
    <interval>5m</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.322.b06-1.el7_9.x86_64/jre/bin</java_path>
    <ciscat_path>/root/Assessor-CLI</ciscat_path>
    <ciscat_binary>Assessor-CLI.sh</ciscat_binary>

    <content type="xccdf" path="benchmarks/CIS_CentOS_Linux_7_Benchmark_v3.1.2-xccdf.xml">
       <profile>"Level 2 - Server"</profile>
    </content>

  </wodle>
```

and the test results (RESTful API and ossec.log) are

<img width="1159" alt="lALPJxDjxECnIo3NA9fNCQ4_2318_983" src="https://user-images.githubusercontent.com/6824093/160772089-41193cb7-bbc2-4230-9382-944dd31b2c9d.png">


Now I can support both CIS-CAT Pro V3 and V4 in both Windows and UNIX systems with wazuh. I would really appreciate if you could review these codes.

vikman90 commented on src/config/wmodules-ciscat.c:
```diff
@@ -172,6 +180,17 @@ int wm_ciscat_read(const OS_XML *xml, xml_node **nodes, wmodule *module)
+            ciscat->ciscat_binary = strdup(nodes[i]->content);
+            #ifdef WIN32
+                if (strcmp(ciscat->ciscat_binary, WM_CISCAT_V3_BINARY_WIN) && strcmp(ciscat->ciscat_binary, WM_CISCAT_V4_BINARY_WIN)) {
+                    mterror(WM_CISCAT_LOGTAG, "Unsupported CIS-CAT Binary '%s'.", ciscat->ciscat_binary);
```
If the configuration is not accepted, the parser should fail.

```suggestion
                    mterror(WM_CISCAT_LOGTAG, "Unsupported CIS-CAT Binary '%s'.", ciscat->ciscat_binary);
                    return OS_INVALID;
```

vikman90 commented on src/config/wmodules-ciscat.c:
```diff
@@ -172,6 +180,17 @@ int wm_ciscat_read(const OS_XML *xml, xml_node **nodes, wmodule *module)
+                }
+            #else
+                if (strcmp(ciscat->ciscat_binary, WM_CISCAT_V3_BINARY) && strcmp(ciscat->ciscat_binary, WM_CISCAT_V4_BINARY)) {
+                    mterror(WM_CISCAT_LOGTAG, "Unsupported CIS-CAT Binary '%s'.", ciscat->ciscat_binary);
```
Same issue here.
```suggestion
                    mterror(WM_CISCAT_LOGTAG, "Unsupported CIS-CAT Binary '%s'.", ciscat->ciscat_binary);
                    return OS_INVALID;
```

vikman90 commented on src/config/wmodules-ciscat.c:
```diff
@@ -172,6 +180,17 @@ int wm_ciscat_read(const OS_XML *xml, xml_node **nodes, wmodule *module)
         } else if (!strcmp(nodes[i]->element, XML_CISCAT_PATH)) {
             ciscat->ciscat_path = strdup(nodes[i]->content);
+        } else if (!strcmp(nodes[i]->element, XML_CISCAT_BINARY)) {
+            ciscat->ciscat_binary = strdup(nodes[i]->content);
```
Free the old string (having the default value) before setting a new one.
```suggestion
            os_free(ciscat->ciscat_binary);
            ciscat->ciscat_binary = strdup(nodes[i]->content);
```

vikman90 commented on src/config/wmodules-ciscat.c:
```diff
@@ -41,6 +42,13 @@ int wm_ciscat_read(const OS_XML *xml, xml_node **nodes, wmodule *module)
     ciscat->scan_config.interval = WM_DEF_INTERVAL;
+    
+    // Set default ciscat binary
+    #ifdef WIN32
+        ciscat->ciscat_binary = WM_CISCAT_V3_BINARY_WIN;
+    #else
+        ciscat->ciscat_binary = WM_CISCAT_V3_BINARY;
+    #endif
```
According to the change below, store these strings in the heap.
```suggestion
    #ifdef WIN32
        os_strdup(WM_CISCAT_V3_BINARY_WIN, ciscat->ciscat_binary);
    #else
        os_strdup(WM_CISCAT_V3_BINARY, ciscat->ciscat_binary);
    #endif
```

vikman90 reviewed (changes requested):
@hustliyilin Thank you very much for your changes, now the code makes much sense!

I have only found an issue in Agentd when rejecting a wrong configuration file. For instance, if we put an invalid value for `<ciscat_binary>` in the manager, at _/var/ossec/etc/shared/default/agent.conf_:

```xml
<agent_config os="linux">
  <wodle name="cis-cat">
    <disabled>no</disabled>
    <timeout>1800</timeout>
    <interval>1d</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>wodles/java</java_path>
    <ciscat_path>wodles/cis-cat</ciscat_path>
    <ciscat_binary>CIS-CAT.sh.wrong</ciscat_binary>
    <content type="xccdf" path="benchmarks/CIS_CentOS_Linux_7_Benchmark_v2.1.1-xccdf.xml">
       <profile>"xccdf_org.cisecurity.benchmarks_profile_Level_2_-_Server"</profile>
    </content>
  </wodle>

</agent_config>
```

The agent shows this log:
```
2022/04/01 12:26:28 wazuh-modulesd:ciscat: ERROR: Unsupported CIS-CAT Binary 'CIS-CAT.sh.wrong'.
```

And it leaks the memory allocated at `ciscat_binary`. I've found that it also leaks `java_path` and `ciscat_path`:

```
==6480== 12 bytes in 1 blocks are definitely lost in loss record 11 of 56
==6480==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
==6480==    by 0x62F1B89: strdup (in /usr/lib64/libc-2.17.so)
==6480==    by 0x4AA66E: wm_ciscat_read (wmodules-ciscat.c:180)
==6480==    by 0x431FE7: Read_WModule (wmodules-config.c:105)
==6480==    by 0x41866A: read_main_elements (config.c:165)
==6480==    by 0x4198C4: ReadConfig (config.c:426)
==6480==    by 0x4333A8: Test_WModule (wmodules-config.c:624)
==6480==    by 0x40BF1B: verifyRemoteConf (restart_agent.c:87)
==6480==    by 0x40CBC8: receive_msg (receiver.c:305)
==6480==    by 0x40A2B7: AgentdStart (agentd.c:209)
==6480==    by 0x40BB38: main (main.c:185)
==6480==
==6480== 12 bytes in 1 blocks are definitely lost in loss record 12 of 56
==6480==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
==6480==    by 0x62F1B89: strdup (in /usr/lib64/libc-2.17.so)
==6480==    by 0x4AA73A: wm_ciscat_read (wmodules-ciscat.c:184)
==6480==    by 0x431FE7: Read_WModule (wmodules-config.c:105)
==6480==    by 0x41866A: read_main_elements (config.c:165)
==6480==    by 0x4198C4: ReadConfig (config.c:426)
==6480==    by 0x4333A8: Test_WModule (wmodules-config.c:624)
==6480==    by 0x40BF1B: verifyRemoteConf (restart_agent.c:87)
==6480==    by 0x40CBC8: receive_msg (receiver.c:305)
==6480==    by 0x40A2B7: AgentdStart (agentd.c:209)
==6480==    by 0x40BB38: main (main.c:185)
==6480==
==6480== 15 bytes in 1 blocks are definitely lost in loss record 17 of 56
==6480==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
==6480==    by 0x62F1B89: strdup (in /usr/lib64/libc-2.17.so)
==6480==    by 0x4AA6D2: wm_ciscat_read (wmodules-ciscat.c:182)
==6480==    by 0x431FE7: Read_WModule (wmodules-config.c:105)
==6480==    by 0x41866A: read_main_elements (config.c:165)
==6480==    by 0x4198C4: ReadConfig (config.c:426)
==6480==    by 0x4333A8: Test_WModule (wmodules-config.c:624)
==6480==    by 0x40BF1B: verifyRemoteConf (restart_agent.c:87)
==6480==    by 0x40CBC8: receive_msg (receiver.c:305)
==6480==    by 0x40A2B7: AgentdStart (agentd.c:209)
==6480==    by 0x40BB38: main (main.c:185)
```

So, I suggest these changes:

1. Free the leaked data at `wm_ciscat_destroy()`, just before freeing `ciscat`:
```c
free(ciscat->java_path);
free(ciscat->ciscat_path);
free(ciscat->ciscat_binary);
```
2. Handle the default values in the heap, as suggested.
3. Free the previous values before assigning new ones.
4. Let the parser reject the configuration if the configuration is wrong, or show a warning and do not set the invalid value to `ciscat_binary`.

I think that we will be ready to merge your PR after applying these changes.

Thank you again!!

Best regards.

hustliyilin commented:
> @hustliyilin Thank you very much for your changes, now the code makes much sense!
> 
> I have only found an issue in Agentd when rejecting a wrong configuration file. For instance, if we put an invalid value for `<ciscat_binary>` in the manager, at _/var/ossec/etc/shared/default/agent.conf_:
> 
> ```
> <agent_config os="linux">
>   <wodle name="cis-cat">
>     <disabled>no</disabled>
>     <timeout>1800</timeout>
>     <interval>1d</interval>
>     <scan-on-start>yes</scan-on-start>
> 
>     <java_path>wodles/java</java_path>
>     <ciscat_path>wodles/cis-cat</ciscat_path>
>     <ciscat_binary>CIS-CAT.sh.wrong</ciscat_binary>
>     <content type="xccdf" path="benchmarks/CIS_CentOS_Linux_7_Benchmark_v2.1.1-xccdf.xml">
>        <profile>"xccdf_org.cisecurity.benchmarks_profile_Level_2_-_Server"</profile>
>     </content>
>   </wodle>
> 
> </agent_config>
> ```
> 
> The agent shows this log:
> 
> ```
> 2022/04/01 12:26:28 wazuh-modulesd:ciscat: ERROR: Unsupported CIS-CAT Binary 'CIS-CAT.sh.wrong'.
> ```
> 
> And it leaks the memory allocated at `ciscat_binary`. I've found that it also leaks `java_path` and `ciscat_path`:
> 
> ```
> ==6480== 12 bytes in 1 blocks are definitely lost in loss record 11 of 56
> ==6480==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
> ==6480==    by 0x62F1B89: strdup (in /usr/lib64/libc-2.17.so)
> ==6480==    by 0x4AA66E: wm_ciscat_read (wmodules-ciscat.c:180)
> ==6480==    by 0x431FE7: Read_WModule (wmodules-config.c:105)
> ==6480==    by 0x41866A: read_main_elements (config.c:165)
> ==6480==    by 0x4198C4: ReadConfig (config.c:426)
> ==6480==    by 0x4333A8: Test_WModule (wmodules-config.c:624)
> ==6480==    by 0x40BF1B: verifyRemoteConf (restart_agent.c:87)
> ==6480==    by 0x40CBC8: receive_msg (receiver.c:305)
> ==6480==    by 0x40A2B7: AgentdStart (agentd.c:209)
> ==6480==    by 0x40BB38: main (main.c:185)
> ==6480==
> ==6480== 12 bytes in 1 blocks are definitely lost in loss record 12 of 56
> ==6480==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
> ==6480==    by 0x62F1B89: strdup (in /usr/lib64/libc-2.17.so)
> ==6480==    by 0x4AA73A: wm_ciscat_read (wmodules-ciscat.c:184)
> ==6480==    by 0x431FE7: Read_WModule (wmodules-config.c:105)
> ==6480==    by 0x41866A: read_main_elements (config.c:165)
> ==6480==    by 0x4198C4: ReadConfig (config.c:426)
> ==6480==    by 0x4333A8: Test_WModule (wmodules-config.c:624)
> ==6480==    by 0x40BF1B: verifyRemoteConf (restart_agent.c:87)
> ==6480==    by 0x40CBC8: receive_msg (receiver.c:305)
> ==6480==    by 0x40A2B7: AgentdStart (agentd.c:209)
> ==6480==    by 0x40BB38: main (main.c:185)
> ==6480==
> ==6480== 15 bytes in 1 blocks are definitely lost in loss record 17 of 56
> ==6480==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
> ==6480==    by 0x62F1B89: strdup (in /usr/lib64/libc-2.17.so)
> ==6480==    by 0x4AA6D2: wm_ciscat_read (wmodules-ciscat.c:182)
> ==6480==    by 0x431FE7: Read_WModule (wmodules-config.c:105)
> ==6480==    by 0x41866A: read_main_elements (config.c:165)
> ==6480==    by 0x4198C4: ReadConfig (config.c:426)
> ==6480==    by 0x4333A8: Test_WModule (wmodules-config.c:624)
> ==6480==    by 0x40BF1B: verifyRemoteConf (restart_agent.c:87)
> ==6480==    by 0x40CBC8: receive_msg (receiver.c:305)
> ==6480==    by 0x40A2B7: AgentdStart (agentd.c:209)
> ==6480==    by 0x40BB38: main (main.c:185)
> ```
> 
> So, I suggest these changes:
> 
> 1. Free the leaked data at `wm_ciscat_destroy()`, just before freeing `ciscat`:
> 
> ```c
> free(ciscat->java_path);
> free(ciscat->ciscat_path);
> free(ciscat->ciscat_binary);
> ```
> 
> 2. Handle the default values in the heap, as suggested.
> 3. Free the previous values before assigning new ones.
> 4. Let the parser reject the configuration if the configuration is wrong, or show a warning and do not set the invalid value to `ciscat_binary`.
> 
> I think that we will be ready to merge your PR after applying these changes.
> 
> Thank you again!!
> 
> Best regards.

Hi @vikman90,
 
Thank you very much. I already modified the codes according to your review and freed the leaked data at `wm_ciscat_destroy()` at https://github.com/wazuh/wazuh/pull/12582/commits/469a1eace6ef87ad09ac953dfe00b59e9389c5fe .  

 I would really appreciate it if you could review these codes again. Thanks again @vikman90, I look forward to your reply.

vikman90 reviewed (approved):
@hustliyilin Thank you for your prompt answer. I've tested your change and now it's OK.

If I fail to define a valid setting, the agent shows this log and it refuses to restart:
```
2022/04/04 09:38:50 wazuh-modulesd:ciscat: ERROR: Unsupported CIS-CAT Binary 'CIS-CAT.sh.wrong'.
2022/04/04 09:38:50 wazuh-agentd: ERROR: (1202): Configuration error at 'etc/shared/agent.conf'.
2022/04/04 09:38:50 wazuh-agentd: ERROR: (1207): WModule remote configuration in 'etc/shared/agent.conf' is corrupted.
```

On the other hand, no memory leaks are found in Agentd:
```
==4757== LEAK SUMMARY:
==4757==    definitely lost: 0 bytes in 0 blocks
==4757==    indirectly lost: 0 bytes in 0 blocks
==4757==      possibly lost: 2,304 bytes in 4 blocks
==4757==    still reachable: 51,890 bytes in 49 blocks
==4757==         suppressed: 0 bytes in 0 blocks
```

We will be ready to merge this PR when we approve the documentation.
Thank you very much !!

hustliyilin commented:
> @hustliyilin Thank you for your prompt answer. I've tested your change and now it's OK.
> 
> If I fail to define a valid setting, the agent shows this log and it refuses to restart:
> 
> ```
> 2022/04/04 09:38:50 wazuh-modulesd:ciscat: ERROR: Unsupported CIS-CAT Binary 'CIS-CAT.sh.wrong'.
> 2022/04/04 09:38:50 wazuh-agentd: ERROR: (1202): Configuration error at 'etc/shared/agent.conf'.
> 2022/04/04 09:38:50 wazuh-agentd: ERROR: (1207): WModule remote configuration in 'etc/shared/agent.conf' is corrupted.
> ```
> 
> On the other hand, no memory leaks are found in Agentd:
> 
> ```
> ==4757== LEAK SUMMARY:
> ==4757==    definitely lost: 0 bytes in 0 blocks
> ==4757==    indirectly lost: 0 bytes in 0 blocks
> ==4757==      possibly lost: 2,304 bytes in 4 blocks
> ==4757==    still reachable: 51,890 bytes in 49 blocks
> ==4757==         suppressed: 0 bytes in 0 blocks
> ```
> 
> We will be ready to merge this PR when we approve the documentation. Thank you very much !!

Hi @vikman90 ,

Thanks very much! Now [the documentation PR](https://github.com/wazuh/wazuh-documentation/pull/4976)  is approved and merged.

I would really appreciate it if you could review and merge this pr again. Thanks again @vikman90, I look forward to your reply.

vikman90 commented:
@hustliyilin we've finally merged this PR! I guess it was closed accidentally, sorry for that.

Thank you very much for your code and patience! We will give credits to you in the change log.

Best!

hustliyilin commented:
> @hustliyilin we've finally merged this PR! I guess it was closed accidentally, sorry for that.
> 
> Thank you very much for your code and patience! We will give credits to you in the change log.
> 
> Best!

Hi @vikman90 ,

Thanks for your help. Wazuh is a great project. I am very willing to continue to contribute issues and codes. And I look forward to working together again. Thanks again.
